Hdu name timeseries branch

williyamshoe opened the pull request:
added hdu name as input to making timeseries. Also added dropdown menus for the hdu and data column names. The dropdown menu is dependent on `timeseries_dropdown.json`, which should be placed in the root of the static folder (in this case, it should be the folder where `qqa monitor` puts all its plots/html files). 
I have not made any code in constructing `timeseries_dropdown.json`, as I was not sure if I should. 

The format for `timeseries_dropdown.json` should look something like this
`{
  "PER_AMP": [
    "READNOISE",
    "COSMICS_RATE",
    "BIAS"
  ],
  "TESTING": [
    "TESTING!!!"
  ]
}`

sbailey commented:
For now let's move `timeseries_dropdown.json` into `py/qqa/cal_files/` and use `pkg_resources.resource_filename` to find it (like you did in `qqa.run.write_tables`).  We can hand-maintain the list of HDU names and QA names, and consider auto-generating it in a separate PR (which could be done by introspecting the contents of `qqa.qa`, which contains the classes that generate those data).

I tried adding a different HDU with a `timeseries_dropdown.json` file like this:
```
{
    "PER_AMP": [ "READNOISE", "COSMICS_RATE", "BIAS" ],
    "PER_CAMERA": [ "MEANDX", "MEANDY" ],
    "TESTING": [ "TESTING!!!" ]
}
```
And it crashed on
```
File "/Users/sbailey/desi/git/qqa/py/qqa/plots/timeseries.py", line 31, in generate_timeseries
list_tables += [fitsio.read(os.path.join(i, file), hdu, columns=["SPECTRO", "CAM", "AMP", "EXPID", "NIGHT", aspect])]
...
ValueError: column name 'AMP' not found (case insensitive)
```
i.e. the `AMP` column only exists in the `PER_AMP` HDU.  `qqa.qa.base.QA.metacols` contains the mapping of which columns are the metadata (i.e. not QA metrics) for each of the HDUs, e.g.
```python
In [4]: from qqa.qa.base import QA

In [5]: QA.metacols
Out[5]: 
{'ALL': ('NIGHT', 'EXPID', 'SPECTRO', 'CAM', 'AMP', 'FIBER'),
 'PER_AMP': ('NIGHT', 'EXPID', 'SPECTRO', 'CAM', 'AMP'),
 'PER_CAMERA': ('NIGHT', 'EXPID', 'SPECTRO', 'CAM'),
 'PER_FIBER': ('NIGHT', 'EXPID', 'SPECTRO', 'FIBER'),
 'PER_CAMFIBER': ('NIGHT', 'EXPID', 'SPECTRO', 'CAM', 'FIBER'),
 'PER_SPECTRO': ('NIGHT', 'EXPID', 'SPECTRO'),
 'PER_EXP': ('NIGHT', 'EXPID')}
```

williyamshoe commented:
In the last commit (1db5c14), I moved `timeseries_dropdown.json` into `py/qqa/cal_files/` and used `pkg_resources.resource_filename` to find it. Also, I think I fixed the missing column(s) issue by only taking the columns displayed on `QA.metacols`, along with the requested qa.status column. However, I  realized that some of these HDUs do not have a `CAM` column, and so I should only be plotting one large timeseries rather than 3 separate ones. Though I can't really test it out on the other HDUs since they are not available yet, I tested it out with the `PER_AMP` HDU by just ignoring the `CAM` column, and it seemed to work fine.

sbailey commented:
Please add an example py/qqa/cal_files/timeseries_dropdown.json file, populated with the HDUs and column names that currently exist.

The dropdowns work well with the specific pages like http://127.0.0.1:5000/timeseries/20200515/20200515/PER_CAMERA/MEANDY ; please also add that to the interface at http://127.0.0.1:5000/timeseries/ .  This should be done by re-using code so that any updates to the usability of one automatically works for the other.

GitHub is also reporting merge conflicts to master; please resolve those (likely easiest via the "resolve conflicts" button on this PR).

Thanks.

sbailey commented:
Looks good.  I removed the CAMFIBER section from the timeseries_dropdown.json config since plotting timeseries for 5k fibers on a single plot takes too long and is too busy.  the PER_AMP and PER_CAMERA are good.  Merging now.

williyamshoe commented:
If that's the case, then I should have it not generate the CAMFIBER section when autogenerating timeseries_dropdown.json as well. I'll make a new branch with this change tomorrow.
